src/dune/fdo.ml

```diff
@@ -0,0 +1,213 @@
+open! Stdune
+module CC = Compilation_context
+
+type phase =
+  | All
+  | Compile
+  | Emit
+
+let linear_ext = ".cmir-linear"
+
+let linear_fdo_ext = linear_ext ^ "-fdo"
+
+let fdo_profile_filename s = s ^ ".fdo-profile"
+
+let linker_script_filename s = s ^ ".linker-script"
+
+let linker_script_hot_filename s = s ^ ".linker-script-hot"
+
+let phase_flags = function
+  | None -> []
+  | Some All -> [ "-g"; "-function-sections" ]
+  | Some Compile ->
+    [ "-g"; "-stop-after"; "scheduling"; "-save-ir-after"; "scheduling" ]
+  | Some Emit -> [ "-g"; "-start-from"; "emit"; "-function-sections" ]
+
+(* CR-soon gyorsh: add a rule to use profile with c/cxx profile if available,
+   similarly to opt_rule for ocaml modules. The profile will have to be
+   generated externally from perf data for example using google's autofdo
+   toolset: create_gcov for gcc or create_llvm_prof for llvm. *)
+let c_flags (ctx : Context.t) =
+  match ctx.fdo_target_exe with
+  | None -> []
+  | Some _ -> [ "-ffunction-sections" ]
+
+let cxx_flags = c_flags
+
+(* Location of ocamlfdo binary tool is independent of the module, but may
+   depend on the context. If it isn't cached elsewhere, we should do it here.
+   CR gyorsh: is it cached? *)
+let ocamlfdo_binary sctx dir =
+  let ocamlfdo =
+    Super_context.resolve_program sctx ~dir ~loc:None "ocamlfdo"
+      ~hint:"try: opam install ocamlfdo"
+  in
+  match ocamlfdo with
+  | Error e -> Action.Prog.Not_found.raise e
+  | Ok _ -> ocamlfdo
+
+(* CR gyorsh: this should also be cached *)
```

Comment:
What's the imperative to cache this? This function seems to be completely pure and seems to do no IO.

Reply:
Ok, I just wanted to make sure we do not repeat `File_tree.file_exists ` for every module we compile, because `fdo_profile` is the same per context.

Reply:
Well `File_tree.file_exists` should be fairly cheap as it's already cached. However, perhaps something should indeed be done.

Some suggestions:

Introduce an `Fdo.Mode` with roughly this signature:

```ocaml
module Mode : sig
  type t = If_exists | Always | Never

  val of_context : Context.t -> t (* or [t option] *)
end
```

Wrap the `profile_exists` and `fdo_profile_src` calculation in a `lazy (...)`.

By the way, regarding performance, I'm now indeed worried that we're running all this code for every module. Perhaps the absence of `OCAMLFDO_USE_PROFILE` should be the same as `Never` and we do this check in `opt_rule` before doing anything else.

Reply:
> Well `File_tree.file_exists` should be fairly cheap as it's already cached.

great! what about `Super_context.resolve_program` - is it also cached? 

> By the way, regarding performance, I'm now indeed worried that we're running all this code for every module. 

These rules are only added for fdo context, so this should not affect normal compilation contexts. 

> Perhaps the absence of OCAMLFDO_USE_PROFILE should be the same as Never and we do this check in opt_rule before doing anything else.

I think the current default of if-exist is a good one from usability point of view. By default, there is no need for the user to set any env var. I think this check can still be done with opt_rule executes.  I'll try to rewrite it.

Reply:
> great! what about Super_context.resolve_program - is it also cached?

Yup. It's cached.

> These rules are only added for fdo context, so this should not affect normal compilation contexts.

Got it. That's good to know.

> I think the current default of if-exist is a good one from usability point of view. By default, there is no need for the user to set any env var. I think this check can still be done with opt_rule executes. I'll try to rewrite it.

By the way, I'm wondering if it should be possible to set this option directly in the workspace file as well.

Reply:
I made some changes to address your suggestions, as far as I understand them.  I haven't been able to test it properly yet, because I am having trouble building packages from opam with the newest dune. What would be the best way to set up an opam switch with a version of dune 2.0 that is based off of an unreleased dune version? 

- Workspace: type of fdo_target_exe is Path.t instead of string.
- Fdo.Mode to read and parse OCAMLFDO_USE_PROFILE env variable from
  context.
- File_tree.file_exists is lazy, and forced only when needed,
  depending on Fdo.Mode.
- Fdo.use_profile is computed dynamically (when opt_rule executes, not
  when it is constructed).
- The result of Fdo.use_profile is now memoized based on context, to
  speed up module compilation under fdo.
- Simplify linker_script_rule: remove linker_script_hot, which is not
  is intended for experiments and can specified via
  OCAMLFDO_LINKER_SCRIPT_FLAGS variable.
- Fix tests.

Reply:
> By the way, I'm wondering if it should be possible to set this option directly in the workspace file as well.

I'm happy to try to add it, perhaps in a separate PR?  
One way is to add another field, like this:
```
(context
 (default
   (fdo src/foo.exe)
   (fdo_use_profile if-exists)))
```
It may be better to have more structure:
```
(context
 (default
   (fdo
    (target_exe src/foo.exe)
    (use_profile if-exists))))
```
There are other fdo-related flags read from environment variables in `fdo.ml`. They could be specified in dune-workspace as well, but I am not sure if it's the right place for all of them. The usage scenarios for them are not entirely clear yet. There could be reasons to specify some of the fdo flags per directory for example.

Reply:
> What would be the best way to set up an opam switch with a version of dune 2.0 that is based off of an unreleased dune version?

`$ opam pin dune --dev` should work I think.

This PR seems good enough for me to merge. One thing that is still needed is a short page in the manual describing this feature.

Also, this is up to your discretion, but you may state that this feature as experimental in the docs. The benefit of this is that you'll be able to break the user facing interface without having to support older versions.

As for the tests, I suppose we should disable them for versions of OCaml less than 4.0.10.

Reply:
> `$ opam pin dune --dev` should work I think.
It did before dune 2.0 (that's how I tested the first version of this PR), but breaks with dune 2.0. I think it has to do with the move of dune-configurator into a separate packaged. I pinned dune-configurator too, and it builds some, but not all packages (after a hack to opam files of dune to convince dune-private-libs and dune-configurator to be friends). I'll try again track it down a bit later today.

> This PR seems good enough for me to merge. 
Great, thank you!

> One thing that is still needed is a short page in the manual describing this feature.
> Also, this is up to your discretion, but you may state that this feature as experimental in the docs. The benefit of this is that you'll be able to break the user facing interface without having to support older versions.
> As for the tests, I suppose we should disable them for versions of OCaml less than 4.0.10.
I'll add these changes shortly, thank you.